enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,482 @@
+node cluster with additional worker nodes, so that my cluster will be able to
+handle larger computation demands.
+
+- As an OpenShift cluster administrator, I want to add worker nodes to my
+existing single control-plane node cluster, so that it'll be able to meet
+growing computation demands.
+
+### API Extensions
+
+This enhancement does not modify/add any API
+
+### Implementation Details/Notes/Constraints
+
+#### Infrastructure Topology value
+
+The following table tries to cover the various installation scenarios, the
+existing values for the "Infrastrcture Topology", the target pool in the
+default `IngressController` and the new proposed values for those two parameters
+under those scenarios.
+
+The following abbreviations are used to get a more compact table -
+
+- `D1W`/`D2W` = The amount of workers during installation / The amount of workers added post-installation
+
+- `SR`/`HA` = SingleReplica / HighlyAvailable
+
+- `CIT`/`PIT` = Current "Infrastrcture Topology" / Proposed "Infrastrcture Topology"
+
+- `M`/`W` = Master / Worker
+
+- `CTP`/`PTP` = Current `IngressController` target pool / Proposed `IngressController` target pool
+
+- `SNO` = Single control-plane node cluster
+
+|Scenario                                |Platform|D1W|D2W|CIT|PIT|CTP|PTP|Is a load balancer currently needed?             |Will a load balancer be needed after proposal?|
+|----------------------------------------|--------|---|---|---|---|---|---|-------------------------------------------------|----------------------------------------------|
+|None platform SNO no workers            |None    |0  |0  |SR |SR |W  |M  |No                                               |No                                            |
+|Cloud platform SNO no workers           |Cloud   |0  |0  |SR |SR |W  |M  |No                                               |No                                            |
+|None platform SNO day1 single worker    |None    |1  |1  |-  |SR |-  |M  |Yes, ingress floats                              |No, ingress pinned to master pool             |
+|Cloud platform SNO day1 single worker   |Cloud   |1  |1  |SR |HA |W  |W  |Yes - provided, ingress floats                   |Yes - provided, there are two ingress replicas|
+|None platform SNO day1 multiple workers |None    |2+ |2+ |-  |SR |-  |M  |Yes, ingress is on two different nodes           |No, one ingress and is pinned to master pool  |
+|Cloud platform SNO day1 multiple workers|Cloud   |2+ |2+ |HA |HA |W  |W  |Yes - provided, ingress is on two different nodes|Yes - provided, there are two ingress replicas|
+|None platform SNO only day2 workers     |None    |0  |1+ |SR |SR |W  |M  |Yes, ingress floats                              |No, ingress pinned to master pool             |
+|Cloud platform SNO only day2 workers    |Cloud   |0  |1+ |SR |SR |W  |M  |Yes - provided, ingress floats                   |Yes - provided, ingress floats                |
```

Comment:
Some questions on this table.....

1) can we denote which of these scenarios are already supported today?  And which "work" today?
2) the first two rows show "zero" workers, but presumably the single node serving as the master is a worker, right?  So it's zero additional workers?  I'm trying to understand how it differentiates from the 3rd and 4th rows which show "one" worker  (Is the master also treated as a worker, when you have a dedicated worker?)
3) It feels like we may be trying to shoehorn the use of the InfrastructureTopology field here.  The fact that we'd have scenarios where we have multiple workers capable of scheduling multiple replicas of infrastructure components, but we will not because of loadbalancer/ingress concerns, makes me think we either need a new field, or a new value for the existing field (e.g. "HighlyAvailableInfraWithSingleNodeIngress" or "HAWithMasterIngress").  Otherwise users looking at the field are likely to be confused by the value it shows.
4) what does the `-` mean in the CIT and CTP columns?  does that mean it's not a scenario that's possible today?
5) in discussing this with @dhellmann we were wondering if maybe even the Cloud scenarios should always be single ingress (targets master).  Yes in theory getting a load balancer is easy, but is it necessary or beneficial, vs the benefits of keeping the SNO experience consistent?  Especially since using SNO is in part a cost decision, and adding the LB in increases cost again.

Reply:
>in discussing this with @dhellmann we were wondering if maybe even the Cloud scenarios should always be single ingress (targets master). Yes in theory getting a load balancer is easy, but is it necessary or beneficial, vs the benefits of keeping the SNO experience consistent? Especially since using SNO is in part a cost decision, and adding the LB in increases cost again.

Defintely +1 from me. I think it should be 1 everywhere. IMO no real need to provide HA solution in such scenarios - if users need ha, they can use ha control plane as a starting point

Reply:
> * can we denote which of these scenarios are already supported today?  And which "work" today?
> * what does the `-` mean in the CIT and CTP columns?  does that mean it's not a scenario that's possible today?

See [this paragraph](https://github.com/openshift/enhancements/blob/278909ed0e513d86445c3320cb928c3a7da819a5/enhancements/single-node/single-node-openshift-with-workers.md?plain=1#L279-L281) about unsupported scenarios. I've also updated the PR with a new column indicating current works/supported/tested for each scenario to clarify this.

> * the first two rows show "zero" workers, but presumably the single node serving as the master is a worker, right?  So it's zero additional workers?  I'm trying to understand how it differentiates from the 3rd and 4th rows which show "one" worker  (Is the master also treated as a worker, when you have a dedicated worker?)

The word "worker" throughout the document refers to nodes which are strictly just workers. So to answer your question - rows 1-2 are just a single node throughout the entire life of the cluster. Rows 3-4 are clusters which get installed with exactly 2 nodes (1 control-plane node and another dedicated worker). Rows 5-6 are clusters which get immediately installed with 3 or more nodes (1 control-plane node and 2 or more dedicated workers). Rows 7-8 are clusters which get installed with just 1 control plane node but then have one or more dedicated worker nodes added to them post-installation.

> * It feels like we may be trying to shoehorn the use of the InfrastructureTopology field here.  The fact that we'd have scenarios where we have multiple workers capable of scheduling multiple replicas of infrastructure components, but we will not because of loadbalancer/ingress concerns, makes me think we either need a new field, or a new value for the existing field (e.g. "HighlyAvailableInfraWithSingleNodeIngress" or "HAWithMasterIngress").  Otherwise users looking at the field are likely to be confused by the value it shows.

Yes I agree with you. I've kinda addressed this [here](https://github.com/openshift/enhancements/blob/9a485478aa068f55d7383393dda0abbecf577800/enhancements/single-node/single-node-openshift-with-workers.md?plain=1#L296-L304). I didn't consider adding a new field / new enum values as a possible solution. It definitely sounds reasonable. I'll think about it

> * in discussing this with @dhellmann we were wondering if maybe even the Cloud scenarios should always be single ingress (targets master).  Yes in theory getting a load balancer is easy, but is it necessary or beneficial, vs the benefits of keeping the SNO experience consistent?  Especially since using SNO is in part a cost decision, and adding the LB in increases cost again.

I want to point out a potential misunderstanding - "getting a load-balancer" is not entirely accurate, a load-balancer is already automatically provided by the IPI installer today - everything there already "just works". Today if you set your install-config.yaml to 1 control plane and n>0 workers on AWS and run the installer, you get a working SNO + workers cluster that behaves correctly even when the ingress floats and you have highly available ingress. Changing the infra topology to SingleReplica in this enhancement will be actively making such deployments worse than they are today for the sake of consistency. Changing the infra topology parameter to SingleReplica would make more sense the day we remove the load balancer from IPI, but with a load-balancer already provided today, we'd just be making things worse. 

Note that I'm not saying that consistency doesn't have a value / is the incorrect decision here, I just want to make sure the situation is well understood before we decide consistency is the way to go. See also [this](https://github.com/openshift/enhancements/pull/996#discussion_r787127690) comment

Reply:
> Today if you set your install-config.yaml to 1 control plane and n>0 workers on AWS and run the installer, you get a working SNO + workers cluster that behaves correctly even when the ingress floats and you have highly available ingress.

but technically this is not an install topology/platform we support today, right?

> Note that I'm not saying that consistency doesn't have a value / is the incorrect decision here

consistency is one piece, cost savings is another piece (keeping SNO's footprint as minimal as possible)

i'm not necessarily trying to push hard for this approach, but i think we should at least make sure we've had the conversation w/ PM about whether it makes sense, for the cloud customers we have in mind for this topology if/when we decide to declare it to be supported

Reply:
>but technically this is not an install topology/platform we support today, right?

That's correct. We do not support it. But it works.

>consistency is one piece, cost savings is another piece (keeping SNO's footprint as minimal as possible)

The point I'm trying to get across is that consistency *is* the only piece

This is because cost saving is out of scope for this enhancement - the load balancer will exist regardless of whether we set the infra topology to HA or not, no money is saved (if I understand correctly, I might have a knowledge gap here then feel free to correct me).

This enhancement does not get rid of the load balancer, so might well keep the infra topology HA.

If future enhancements want to do away with the load balancer in single control-plane node clusters in the cloud for whatever reason - those enhancement should re-define the value of the infra topology field to be SingleReplica.

Reply:
> This is because cost saving is out of scope for this enhancement - the load balancer will exist regardless of whether we set the infra topology to HA or not, no money is saved

thanks, I did not realize it was being created regardless in this case.

At a minimum i think we need to ensure that answering the question of what LB behavior we want, is defined before we declare support for SNO on cloud.  That's outside the scope of this EP, but probably needs to be called out in this EP.  Removing the cloud entries from the table (as i think was suggested elsewhere) might help to reduce the confusion.  Or at least making it clear that the behavior defined for cloud entries is being described as "how it works today, but it's not supported and we would consider changing how it works before making it a supported topology"